src/Nri/Ui/Util.elm

```diff
@@ -16,13 +16,74 @@ dashify =
 
 
 {-| Convenience method for removing punctuation
-(removes everything that isn't whitespace or alphanumeric).
+(removes everything that isn't whitespace, alphanumeric, or an underscore).
 -}
 removePunctuation : String -> String
 removePunctuation =
     let
         regex =
-            Regex.fromString "[^A-z0-9\\w\\s]"
+            Regex.fromString "[^A-Za-z0-9\\w\\s]"
```

Comment:
"But Alex, why is `[A-Za-z]` different from `[A-z]` when the majuscule letters precede the minuscules in ASCII (and thus UTF-8) character encoding?? Surely these are the same"

No, young learner. They are not. For in the before time, in the long long ago, we decided to be *clever*. And endowed as we are with a mere 26 letters in English and 32 being a nice round number in binary representation<sup>1</sup>, we decided that for any majuscule character stored in memory as a number, we could *idempotently* find its minuscule by performing a bitwise OR with the number 32. To do this we would represent "A" with 65 and "a" with 97 and pad out the middle with the square brackets, backslash, underscore, and the backpack.  And of course character ranges in regexes respect that and match only those punctuation marks in between Z and a.

But let's turn to some other troubling bits of this: `\w` means "any word character" which is a bit rich. It contains `A-Z` and `a-z` and `0-9` and `_` to boot. But it also matches just about any roughly Latin alphabet word character according to a number of regex testers (a fair bit short of *any* word character I'd say) but that's not even real life. Here's me trying to make a URL slug or an ID for a French seasonal fashion line:

```
> summer = "éte"+String.fromCharCode(0x0301)
'été'
> for(let i = 0, c = -1; !isNaN(c); ++i ) { c =  summer.charCodeAt(i); console.log(i, c)}
0 233
1 116
2 101
3 769
4 NaN
undefined
> summer.replace(/[^\w]/g, '-') // replace every non-word character with a hyphen-minus
'-te-'
> summer.replace(/\W/g, '-') // replace every non-word character with a hyphen-minus the other way
'-te-'
```

So really this regex could just be `Regex.fromString "[\\W\\S]"` but I'd like to know the author's intent first. 

---

1. Were that the soviets had won out with their nice 32 character alphabet in Russian<sup>2</sup>
2. yes I know the other Cyrillic languages use more letters<sup>3</sup>
3. yes I know that the ternary [Setun](https://en.wikipedia.org/wiki/Setun) may have made this a moot point)

Reply:
@tesk9 @erik-nri you're on a ride with me now. This isn't used in the monolith. Rather, in the monolith there's `ui/src/Util` which exports `removePunctuation` which is used only in `ui/src/Page/Learn/PeerReviews/Rate/Paragraph/BinaryJustificationStep.elm` in the function `generateId` which should PROBABLY just use our new safeIdString. Anyone wanna hackday with me on that tomorrow?

 Also remember when I said "I'd like to know the author's intent first" well the author is *in the room right now!*
https://github.com/NoRedInk/NoRedInk/commit/d712227a1a7c0d9e05d742b164fab556b07c655b

Reply:
HAHAHAHAH we are laughing we are rejoicing we are having the very best time.

Reply:
😆 the author isn't super good at regex! safeIdString for everybody everywhere!